### PR TITLE
Filter Implementation

### LINKED ISSUE #9: Filter Class
As a daft engineer, I want to create real time bandpass filters in C++ in order to seperate freqeuncy components of the human voice for vocoding.

Success Criteria:
- UML Class Diagram
- Research filter types (likely IIR)
- Correct sample length / buffer size for each event

### PR DESCRIPTION
**Pull request by MarkAHarley**
Closes #9 

Added a filter class which generates an IIR bandpass with given cutoffs and supports threading and message passing.

### CODE REVIEW

File: include/filter/filter.hh [hunk 1]
```diff
@@ -1,4 +1,35 @@
...
+    Audio filter(const Audio &in_audio);
+
+  private:
+    static const int default_order{12};
+    Iir::Butterworth::BandPass<default_order> f;
+    double centre_freq{0};
+    double freq_range{UINT_MAX};
+
+    Pipe<Audio> *input;
+    Pipe<Audio> *output;
```
Comment (MHCooke) [inline include/filter/filter.hh:29]: Can you make these references?

File: include/filter/filter.hh [hunk 2]
```diff
@@ -1,4 +1,35 @@
...
 
-#endif // FILTER_FILTER_HH
+#include "../audio/audio.hh"
+#include "../pipe.hh"
+#include <Iir.h>
+#include <climits>
+#include <condition_variable>
+#include <cstdio>
+#include <memory>
+#include <thread>
```
Comment (MHCooke) [inline include/filter/filter.hh:11]: are climits, cstdio, and memory still required?

**Review by MHCooke — COMMENTED** [review]
Looks pretty good in general. Could you add new lines at the end of your files (you can probably set clion to do that)

Other than that the reference instead of pointers stuff is my only real suggestion

File: unit_tests/filter_test.cc [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    }};
+
+    std::cerr << "Exiting" << std::endl;
+    f.stop();
+    std::cerr << "Filter stopped" << std::endl;
+    input_thread.join();
+    std::cerr << "Input stopped" << std::endl;
+    output_thread.join();
+    std::cerr << "Output stopped" << std::endl;
+}
```
Comment (MHCooke) [inline unit_tests/filter_test.cc:197]: I'm not convinced this unit test doesn't suffer from the same hanging problem mine did. Can you add synchronisation logic to ensure that the filter will only every try to loop once.

File: include/filter/filter.hh (hunk 2 above)
Comment (MarkAHarley) [reply include/filter/filter.hh:11]: climits is used for default filter values, but could remove the other two.

File: src/filter.cc [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    filter_thread = std::thread([this]() {
+        std::unique_lock<std::mutex> lk(input.cond_m);
+        running = true;
+        thread_alive = true;
+        while (running) {
+            input.cond.wait(lk, [this] { return input.queue.empty() == false; });
+
+            Audio audio_out = filter(input.queue.front());
+            input.queue.pop();
+
```
Comment (Jambls) [inline src/filter.cc:26]: You should unlock your input pipe when you're not using it, unless the filter running is really quick. lk.unlock(); and lk.lock(); should do the trick. See #41 for more.

**Review by Jambls — COMMENTED** [review]
Looks good other than that

**Review by bear-in-the-air — APPROVED** [review]

File: src/filter.cc (hunk 4 above)
Comment (MarkAHarley) [reply src/filter.cc:29]: Done

Comment (Jambls) [reply src/filter.cc:29]: epic

**Review by Jambls — APPROVED** [review]

**Comment by Jambls** [comment]
You've sorted matthew's comments right? Should be good to go.

**Comment by MarkAHarley** [comment]
> You've sorted matthew's comments right? Should be good to go.

Yep all sorted

**Pull request merged**